client/src/components/TreeDropdownField/TreeDropdownField.js

```diff
@@ -20,38 +23,51 @@ class TreeDropdownField extends Component {
     this.render = this.render.bind(this);
     this.renderMenu = this.renderMenu.bind(this);
     this.renderOption = this.renderOption.bind(this);
+
     // Getters
     this.getBreadcrumbs = this.getBreadcrumbs.bind(this);
     this.getDropdownOptions = this.getDropdownOptions.bind(this);
     this.getSelectedOption = this.getSelectedOption.bind(this);
     this.getVisibleTree = this.getVisibleTree.bind(this);
+
     // Events
     this.handleBack = this.handleBack.bind(this);
     this.handleChange = this.handleChange.bind(this);
     this.handleKeyDown = this.handleKeyDown.bind(this);
     this.handleNavigate = this.handleNavigate.bind(this);
+    this.handleSearchChange = this.handleSearchChange.bind(this);
+    this.handleSearchReset = this.handleSearchReset.bind(this);
+
     // Helpers
     this.callFetch = this.callFetch.bind(this);
     this.lazyLoad = this.lazyLoad.bind(this);
     this.findTreeByID = this.findTreeByID.bind(this);
     this.findTreeByPath = this.findTreeByPath.bind(this);
     this.findTreePath = this.findTreePath.bind(this);
+
+    this.searchTimer = null;
   }
 
   componentDidMount() {
     // Ensure root node is loaded, force invalidating the cache when not readonly or disabled
     if (!this.props.readOnly && !this.props.disabled) {
-      this.loadTree([]);
+      this.loadTree([], true, this.props.search);
     }
   }
 
   componentWillReceiveProps(nextProps) {
-    if (nextProps.data.cacheKey !== this.props.data.cacheKey
-      && !this.props.readOnly
-      && !this.props.disabled
-    ) {
+    if (this.props.readOnly || this.props.disabled) {
+      return;
+    }
+
+    if (nextProps.data.cacheKey !== this.props.data.cacheKey) {
       // invalidate the tree cache, as paths have changed
-      this.loadTree([]);
+      this.loadTree([], false, nextProps.search);
+    }
+
+    if (this.props.search !== nextProps.search) {
```

Comment:
no, I don't think it's possible in an application setting.
the `cacheKey` comes from schema, so it'll be two separate redux update calls and it'd only be only or the other updated and unlikely both at the same time.

just thinking about the order, perhaps it could be safer with `(this.props.search !== nextProps.search)` being checked first, but not necessary :)